3_time_series/Statistical_analysis.md

```diff
@@ -0,0 +1,749 @@
+        model_fit.append(fit)
+        predictions.append(pred)
+    
+    predictions_df = pd.DataFrame(pd.concat(predictions, axis=0), columns=['Num_Passengers'])
+    return model_fit, predictions_df
+```
+
+```python
+def fit_model_sarimax(order, seasonal_order, dataframe, split, verbose=False):
+    model_fit = []
+    predictions =  []
+    
+    for i, cur_date in enumerate(pd.date_range(split,  dataframe.index[-1], freq='MS')):
+        if(verbose): print(i, " iteration")
+        model = SARIMAX(dataframe[:cur_date - timedelta(days=1)], order=order, seasonal_order=seasonal_order)
+        fit = model.fit()
+        pred = fit.predict(start=cur_date, end=cur_date) 
+        
+        model_fit.append(fit)
+        predictions.append(pred)
+    
+    predictions_df = pd.DataFrame(pd.concat(predictions, axis=0), columns=['Num_Passengers'])
+    return model_fit, predictions_df
+```
+
+```python
+def get_split_date(data, split_coef):
+    if split_coef<0 or split_coef>1:
+        raise ValueError("Split coefficient should be in range [0,1]")
+    split = math.floor(len(data) * split_coef)
+    split = split if split<len(data) else split-1
+    return data.index[split]
+```
+
+```python
+def split_train_test(df_to_split, train_percents):
+    split_date = get_split_date(df_to_split, train_percents)
+
+    train_data = df_to_split[:split_date - timedelta(days=1)]
+    test_data = df_to_split[split_date:]
+    return train_data, test_data, split_date
+```
+
+```python
+def split_train_test_by_date(df_to_split, split_date):
+    train_data = df_to_split[:split_date - timedelta(days=1)]
+    test_data = df_to_split[split_date:]
+    return train_data, test_data
+```
+
+```python
+def get_metrics_for_df(test_data, predictions, col='Num_Passengers'):
+    return get_metrics(test_data[col], predictions[col])
+```
+
+```python
+def get_metrics(test_data, predictions):
+    mse = mean_squared_error(test_data, predictions)  
+    rmse = mean_squared_error(test_data, predictions, squared=False)          
+    mae = mean_absolute_error(test_data, predictions)
+    return mse, rmse, mae
+```
+
+```python
+def print_metrics(mse, rmse, mae):
+    print("MSE - ", mse)
+    print("RMSE - ", rmse)
+    print("MAE - ", mae)
+```
+
+```python
+def plot_test_with_predictions(test, pred, name):
+    plt.figure(figsize=(8,4))
+
+    plt.plot(test)
+    plt.plot(pred)
+    plt.title(name)
+    plt.legend(('Data', 'Predictions'), fontsize=16)
+    plt.show()
+```
+
+```python
+def inverse_diffirence(observation, diffs, periods):
+    restored = observation.copy()
+    restored.iloc[periods+1:] = np.nan
+    for d, val in diffs.iterrows():
+        restored.loc[d] = restored.loc[d - pd.DateOffset(months=periods)] + val
+    return restored.iloc[periods:]
+```
+
+<!-- #region tags=[] -->
+## Baseline
+<!-- #endregion -->
+
+<!-- #region tags=[] -->
+### Predict previous value
+<!-- #endregion -->
+
+```python
+_, b_test_data, split_date = split_train_test(df, 0.70)
```

Comment:
well we have to separate fitting and test data, as far as I understand right now we are fitting on whole data and testing on last 30% of the values (on which we fitted)